### PR TITLE
ConstraintElim: teach fact-transfer about samesign

### PR DESCRIPTION
**Pull request by artagnon**
When the samesign flag is present on an icmp, we can transfer all the facts on the unsigned system to the signed system, and vice-versa: we do this by specializing transferToOtherSystem when samesign is present.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 1]
```diff
@@ -1794,7 +1803,14 @@ static bool eliminateConstraints(Function &F, DominatorTree &DT, LoopInfo &LI,
       if (ReproducerModule && DFSInStack.size() > ReproducerCondStack.size())
         ReproducerCondStack.emplace_back(Pred, A, B);
 
-      Info.transferToOtherSystem(Pred, A, B, CB.NumIn, CB.NumOut, DFSInStack);
+      // If samesign is present on the ICmp, simply transfer the signed system
+      // to the unsigned system, and viceversa.
+      if (HasSameSign)
+        Info.addFact(CmpInst::getFlippedSignednessPredicate(Pred), A, B,
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1809]: Should avoid duplicating facts with equality predicates.

Comment (artagnon) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1809]: I would have expected samesign to be illegal on ICmps with equality predicates? What sense does `icmp samesign eq` or `icmp samesign ne` make?

Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1809]: It is allowed by spec.

**Review by dtcxzyw — APPROVED** [review]
LG. Please wait for additional approval from other reviewers.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 2]
```diff
@@ -1794,7 +1803,17 @@ static bool eliminateConstraints(Function &F, DominatorTree &DT, LoopInfo &LI,
       if (ReproducerModule && DFSInStack.size() > ReproducerCondStack.size())
         ReproducerCondStack.emplace_back(Pred, A, B);
 
-      Info.transferToOtherSystem(Pred, A, B, CB.NumIn, CB.NumOut, DFSInStack);
+      if (ICmpInst::isRelational(Pred)) {
+        // If samesign is present on the ICmp, simply transfer the signed system
+        // to the unsigned system, and viceversa.
```
Comment (antoniofrighetto) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1808]:
```suggestion
        // If samesign is present on the ICmp, simply flip sign predicate and transfer such an information from the signed system
        // to the unsigned system, and viceversa.
```

**Review by nikic — COMMENTED** [review]
Probably makes sense to land https://github.com/llvm/llvm-project/pull/116867 first and then use it here...

**Comment by artagnon** [comment]
> Probably makes sense to land #116867 first and then use it here...

I realized that it's also better to first land #118534.

**Comment by artagnon** [comment]
Rebased for a clean implementation.

**Comment by dtcxzyw** [comment]
Ping. Is it ready to land?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**